SimplenoteWidgets/Tools/WidgetDataController.swift

```diff
@@ -0,0 +1,100 @@
+import Foundation
+import SimplenoteFoundation
+import SimplenoteSearch
+
+class WidgetDataController {
+
+    /// Data Controller
+    ///
+    let coreDataManager: CoreDataManager
+
+    /// Notes Controller
+    ///
+    private lazy var notesController = ResultsController<Note>(
+        viewContext: coreDataManager.managedObjectContext,
+        matching: predicateForNotes(),
+        sortedBy: [NSSortDescriptor.descriptorForNotes(sortMode: noteSortMode)]
+    )
+
+    /// Initialization
+    ///
+    init(coreDataManager: CoreDataManager, isPreview: Bool = false) throws {
+        guard !isPreview else {
+            self.coreDataManager = coreDataManager
+            return
+        }
+
+        guard let isLoggedIn = UserDefaults(suiteName: SimplenoteConstants.sharedGroupDomain)?.bool(forKey: .accountIsLoggedIn),
+              isLoggedIn else {
+            throw WidgetError.appConfigurationError
+        }
+
+        self.coreDataManager = coreDataManager
```

Comment:
There's a bit of duplication: `self.coreDataManager = coreDataManager` is implemented twice
One alternative could be:


```
 init(coreDataManager: CoreDataManager, isPreview: Bool = false) throws {
    self.coreDataManager = coreDataManager

    if isPreview {
        return
    }

    let authenticated = UserDefaults(suiteName: SimplenoteConstants.sharedGroupDomain)?.bool(forKey: .accountIsLoggedIn) ?? false

    guard authenticated else {
        throw WidgetError.appConfigurationError
        return
    }

```

Reply:
So.... The problem here is SwiftUI previews.  SwiftUI previews have no access to UserDefaults, so by adding in this check to make sure a user is logged in, it breaks the SwiftUI previews.  The first guard to see if it is a preview is meant to exit with a setup WidgetDataController without checking if the user is logged in.  I this solution above  we would still setup the WidgetDataController if the user is logged out, but would throw an error.  I'm not sure that is what we want here, but this solution works

```
    init(coreDataManager: CoreDataManager, isPreview: Bool = false) throws {
        if !isPreview {
            guard let isLoggedIn = UserDefaults(suiteName: SimplenoteConstants.sharedGroupDomain)?.bool(forKey: .accountIsLoggedIn),
                  isLoggedIn else {
                throw WidgetError.appConfigurationError
            }
        }

        self.coreDataManager = coreDataManager
    }
```

The other solution is something I have been toying with, and that is removing SwiftUI previews from the widgets before it is merged.

Reply:
Gotcha!! thanks for changing that!!